src/algebra/module_regular.lean

```diff
@@ -0,0 +1,167 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.group
+import algebra.group_power.basic
+import algebra.module.basic
+import algebra.smul_with_zero
+/-!
+# Action of regular elements on a module
+
+We introduce `M`-regular elements, in the context of an `R`-module `M`.
+
+By definition, an `M`-regular element of a commutative ring acting on an `R`-module `M` is an
+element `a ∈ R` such that the function `m ↦ a • m` is injective.
+-/
+
+--should it be in the module namespace?
+namespace module
+
+variables {R S : Type*} {M : Type*} {a b : R} {s : S}
+
+section has_scalar
+
+variables [has_scalar R M] [has_scalar R S] [has_scalar S M] [is_scalar_tower R S M]
+
+variables (M)
+
+/-- An `M`-regular element is an element `c` such that multiplication on the left by `c` is an
+injective map `M → M`. -/
+def is_regular (c : R) := function.injective ((•) c : M → M)
+
+variables {M}
+
+/-- In a monoid, the product of `M`-regular elements is `M`-regular. -/
+lemma is_regular.smul (ra : is_regular M a) (rs : is_regular M s) :
+  is_regular M (a • s) :=
+λ a b ab, rs (ra ((smul_assoc _ _ _).symm.trans (ab.trans (smul_assoc _ _ _))))
+
+/--  If an element `b` becomes `M`-regular after multiplying it on the left by an `M`-regular
+element, then `b` is `M`-regular. -/
+lemma is_regular.of_smul (a : R) (ab : is_regular M (a • s)) :
+  is_regular M s :=
+@function.injective.of_comp _ _ _ (λ m : M, a • m) _ (λ c d cd, ab
+  (by rwa [smul_assoc, smul_assoc]))
+
+/--  An element is `M`-regular if and only if multiplying it on the left by an `M`-regular element
+is `M`-regular. -/
+@[simp] lemma is_regular.smul_iff (b : S) (ha : is_regular M a) :
+  is_regular M (a • b) ↔ is_regular M b :=
+⟨is_regular.of_smul _, ha.smul⟩
+
+end has_scalar
+
+section monoid
+
+variables [monoid R] [mul_action R M]
+
+/--  Two elements `a` and `b` are `M`-regular if and only if both products `a • b` and `b • a`
+are `M`-regular. -/
+lemma is_regular_mul_and_mul_iff :
+  is_regular M (a * b) ∧ is_regular M (b * a) ↔ is_regular M a ∧ is_regular M b :=
+begin
+  refine ⟨_, _⟩,
+  { rintros ⟨ab, ba⟩,
+    exact ⟨ba.of_smul b, ab.of_smul a⟩ },
+  { rintros ⟨ha, hb⟩,
+    exact ⟨(ha.smul_iff b).mpr hb, (hb.smul_iff a).mpr ha⟩ }
+end
+
+/--  The "most used" implication of `mul_and_mul_iff`, with split hypotheses, instead of `∧`. -/
+lemma is_regular.and_of_mul_of_mul (ab : is_regular M (a * b)) (ba : is_regular M (b * a)) :
+  is_regular M a ∧ is_regular M b :=
+⟨ba.of_smul b, ab.of_smul a⟩
```

Comment:
Sorry, to be clear, my point was: to remove this lemma entirely
```suggestion
end
```

Reply:
I want to think about this a little: in the "algebra" case, with no actions, this is used quite a lot, so I am even tempted to make it a simp lemma.  In this case, I have not thought about it enough to really know for sure.  I will hold onto this for little bit and will let you know!

Reply:
Are you sure this is used quite a lot? I'm envisaging you using:
```lean
obtain ⟨ra, rb⟩ := ab.and_of_mul_of_mul ba
```
Is that really better than obtaining the two pieces separately?
```lean
obtain ra := ba.of_smul a,
obtain rb := ab.of_smul b,
```

Reply:
Ok, I am convinced and I removed this lemma!